### PR TITLE
Implement a HttpService backend that runs on the AWS Lambda Rust Runtime

### PR DESCRIPTION
**Pull request by kanru**
## Description

Add a new server backend that runs on the `aws-lambda-rust-runtime`. Example usage:

```rust
fn main() {
    let mut app = tide::App::new(());
    app.at("/").get(async move |_| "Hello, world!");
    http_service_lambda::run(app.into_http_service());
}
```

## Motivation and Context

Because I wanted to use a framework like `tide` to handle requests in a Lambda function.
Related issue: https://github.com/rustasync/http-service/issues/10

## How Has This Been Tested?

Build the example code and deployed to a Lambda function. I also ported one app that uses tide to Lambda.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/rust-net-web/tide/blob/master/.github/CONTRIBUTING.md) document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: http-service-lambda/Cargo.toml
Comment (secretfader) [inline http-service-lambda/Cargo.toml]: IMO, it makes sense to use `authors` from `http-service-hyper` and other crates here, too. (Also: I'd apply for membership in the GitHub organization.)

File: http-service-lambda/Cargo.toml
Comment (secretfader) [inline http-service-lambda/Cargo.toml]: It's common for components that rely on Tide to [depend on Git `master` in development](https://github.com/rustasync/tide/blob/master/tide-compression/Cargo.toml#L30). We might want to adopt that practice here as well.

**Comment by secretfader** [comment]
I made a few comments in-line on this PR, but otherwise it looks solid and gets my approval. :+1:

File: http-service-lambda/Cargo.toml
Comment (kanru) [reply http-service-lambda/Cargo.toml]: Hi, I tried to implement this but I think I don't fully understand what you mean. It seems each crate has different `authors` set and it doesn't make sense to make a union of them here.

I'd love to apply for membership. Where/who can I put such application?

File: http-service-lambda/Cargo.toml
Comment (kanru) [reply http-service-lambda/Cargo.toml]: This sounds a good idea, but I'm not sure how to solve the `http-service` dependency. Currently I have

```toml
[dependencies]
http-service = "0.2.0"
```

But I would like to use

```toml
[dependencies]
http-service = { version = "0.2.0", path = ".." }
```

It creates a conflict between Tide because Tide wants to use the git version instead of the local version:

```sh
% cargo tree -i -p https://github.com/rust-lang/crates.io-index#http-service:0.2.0
http-service v0.2.0
├── http-service-hyper v0.2.0
│   └── tide v0.2.0 (https://github.com/rustasync/tide#25de4138)
│       [dev-dependencies]
│       └── http-service-lambda v0.1.0 (/home/kanru/r/http-service/http-service-lambda)

% cargo tree -i -p file:///home/kanru/r/http-service#0.2.0
http-service v0.2.0 (/home/kanru/r/http-service)
└── http-service-lambda v0.1.0 (/home/kanru/r/http-service/http-service-lambda)
```

Comment (secretfader) [reply http-service-lambda/Cargo.toml]: Sounds like an issue we may need to patch up elsewhere. Thanks for sharing the problem you encountered! I'm not sure how to resolve this error either, but I'll do some research and see if a workable solution can be found.

File: http-service-lambda/Cargo.toml
Comment (secretfader) [reply http-service-lambda/Cargo.toml]: Sorry I'm just now getting back to this.

[Here's an example of how the `authors` line in your `Cargo.toml` should look](https://github.com/rustasync/http-service/blob/master/http-service-hyper/Cargo.toml#L2-L5). It's cool to substitute your name and email for the last entry in my example, but I think it makes sense for maintainers to be able to publish this package alongside others in the repo.

As for applying for rustasync membership, you can ask in Discord (link on the [org page](https://github.com/rustasync)). I don't hang out much there anymore, but plenty of other folks do.

File: http-service-lambda/Cargo.toml
Comment (secretfader) [reply http-service-lambda/Cargo.toml]: Today I learned this has to be overridden in the main `Cargo.toml` for a workspace. Leave this as is, it's possible to update to better practices in a separate PR.

**Comment by secretfader** [comment]
I pushed a change that resolves dependencies in this repo via path, and Tide via Git. It seems to have exposed one place where the API wasn't quite up-to-date with `master`, and I resolved that.

All in all, I think this is a really solid contribution and I'm glad you made it. We can hopefully get this merged shortly.

**Comment by secretfader** [comment]
@rustasync/maintainers LGTM. Could someone else review before merging?

File: http-service-lambda/Cargo.toml
Comment (secretfader) [reply http-service-lambda/Cargo.toml]: I went ahead and resolved this while I was working on other things this evening. Hope you don't mind.

Comment (kanru) [reply http-service-lambda/Cargo.toml]: No problem. Thanks for updating the PR!

File: http-service-lambda/src/lib.rs
Comment (Nemo157) [inline http-service-lambda/src/lib.rs]: Not a big deal since it's a private trait, but spelling mistake in this trait name

```suggestion
trait CompatHttpBodyAsLambda {
```

Comment (kanru) [reply http-service-lambda/src/lib.rs]: Good catch!

File: http-service-lambda/src/lib.rs
Comment (Nemo157) [inline http-service-lambda/src/lib.rs:60]: This makes me uncomfortable, what if it is binary data which just happens to be utf-8 compatible? Do the different body types affect how the API gateway returns the bodies to the client? Or does it just affect the base64 encoding as it passes the value from here to the gateway?

**Comment by Nemo157** [comment]
Awesome to see more `http-service` implementations, just made a couple small comments, really don't care about the spelling issue, but the distinction between text and binary data makes me wonder if there's an issue in having `http-service` require all bodies to be `Vec<u8>` equivalents instead of allowing more structured bodies through.

File: http-service-lambda/src/lib.rs
Comment (kanru) [reply http-service-lambda/src/lib.rs:60]: This is a good question. For API gateway, it doesn't matter very much because the API gateway will do the final conversion: https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-payload-encodings-workflow.html so the check here only save us one extra serialization to base64 and back.

For ELB, a base64 encoded body is required to return binary data: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#respond-to-load-balancer, but I think in the end it's the `content-type` header decides how the client should handle the response.

Technically we can always return base64 encoded data, but we can save the serialization effort when we don't need to.

**Review by Nemo157 — APPROVED** [review]

**Comment by secretfader** [comment]
It does seem like we should explore the topic of structured bodies. That will probably be useful conversation if we aim to support other protocols, like gRPC, in the future.

Thanks, @nemo157 for catching that.

**Pull request merged**
